crates/sqllib/src/string.rs

```diff
@@ -47,6 +49,58 @@ impl SqlString {
         SqlString(StringRef::from(value.to_string()))
     }
 
+    /// Constructs a `SqlString` that contains `value`.
+    ///
+    /// If `candidate` is the same as `value`, then reuses it instead of
+    /// allocating a new string and copying into it.
+    pub fn maybe_reuse(value: &str, candidate: &SqlString) -> Self {
+        if value.as_ptr() == candidate.str().as_ptr() && value.len() == candidate.len() {
+            candidate.clone()
+        } else {
+            SqlString(StringRef::from(value.to_string()))
+        }
+    }
+
+    /// Constructs a `SqlString` that contains all of `strings` concatenated.
+    ///
+    /// This is more efficient than concatenating into a `String` and then
+    /// converting it to a `SqlString` because `ArcStr::from` always allocates
+    /// and copies.
+    pub fn from_concat(strings: &[&str]) -> Self {
+        Self::from_concat_iterator(strings.iter().copied())
+    }
+
+    /// Constructs a `SqlString` that contains the concatenation of the strings
+    /// produced by the iterator.
+    ///
+    /// The iterator gets iterated twice and must produce the same total length
+    /// (normally, the same strings) each time.
+    pub fn from_concat_iterator<'a, I>(strings: I) -> Self
+    where
+        I: IntoIterator<Item = &'a str>,
+        I::IntoIter: Clone,
+    {
+        fn copy_bytes_uninit(dst: &mut [MaybeUninit<u8>], src: &[u8]) {
```

Comment:
I promised to never use unsafe code. I am glad you will do the dirty jobs for me